app/src/main/java/at/bitfire/icsdroid/model/SubscriptionsModel.kt

```diff
@@ -240,27 +259,50 @@ class SubscriptionsModel(application: Application): AndroidViewModel(application
                 // sync the subscription to reflect the changes in the calendar provider
                 SyncWorker.run(getApplication())
 
-                withContext(Dispatchers.Main) {
-                    toast.cancel()
-                    Toast.makeText(
-                        context,
-                        context.resources.getQuantityString(R.plurals.backup_imported, newSubscriptions.size, newSubscriptions.size),
-                        Toast.LENGTH_SHORT
-                    ).show()
-                }
+                toastAsync(
+                    message = {
+                        resources.getQuantityString(R.plurals.backup_imported, newSubscriptions.size, newSubscriptions.size)
+                    },
+                    cancelToast = toast
+                )
             } catch (e: JSONException) {
                 Log.e(TAG, "Could not load JSON: $e")
-                withContext(Dispatchers.Main) {
-                    toast.cancel()
-                    Toast.makeText(context, R.string.backup_import_error_json, Toast.LENGTH_LONG).show()
-                }
+                toastAsync(
+                    messageResId = R.string.backup_import_error_json,
+                    cancelToast = toast,
+                    duration = Toast.LENGTH_LONG
+                )
             } catch (e: SecurityException) {
                 Log.e(TAG, "Could not load JSON: $e")
-                withContext(Dispatchers.Main) {
-                    toast.cancel()
-                    Toast.makeText(context, R.string.backup_import_error_security, Toast.LENGTH_LONG).show()
-                }
+                toastAsync(
+                    messageResId = R.string.backup_import_error_security,
+                    cancelToast = toast,
+                    duration = Toast.LENGTH_LONG
+                )
+            } catch (e: IOException) {
+                Log.e(TAG, "Could not load JSON: $e")
+                toastAsync(
+                    messageResId = R.string.backup_import_error_io,
+                    cancelToast = toast,
+                    duration = Toast.LENGTH_LONG
+                )
             }
         }
     }
+
+    private suspend fun toastAsync(
+        message: (Context.() -> String)? = null,
+        @StringRes messageResId: Int? = null,
+        cancelToast: Toast? = null,
+        duration: Int = Toast.LENGTH_SHORT
+    ): Toast = withContext(Dispatchers.Main) {
+        cancelToast?.cancel()
+        val context: Context = getApplication()
+
+        when {
+            message != null -> Toast.makeText(context, message(context), duration)
+            messageResId != null -> Toast.makeText(context, messageResId, duration)
+            else -> throw IllegalArgumentException("Either message or messageResId must be given.")
```

Comment:
Looks good. And good thinking to extract the toast logic, but throwing the exception here is a bit extreme. How about just logging and not showing the toast message? I feel we might be setting ourselves up for some trouble otherwise. :)